org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend

```diff
@@ -0,0 +1,47 @@
+/*******************************************************************************
+ * Copyright (c) 2018 TypeFox and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.ide.server.symbol
+
+import com.google.inject.ImplementedBy
+import java.util.List
+import org.eclipse.lsp4j.DocumentSymbol
+import org.eclipse.lsp4j.DocumentSymbolParams
+import org.eclipse.lsp4j.SymbolInformation
+import org.eclipse.lsp4j.jsonrpc.messages.Either
+import org.eclipse.xtext.ide.server.Document
+import org.eclipse.xtext.resource.XtextResource
+import org.eclipse.xtext.util.CancelIndicator
+
+/**
+ * Common service interface for providing document symbol information for text documents.
+ * 
+ * <p>
+ * For more details, see the <a href="https://microsoft.github.io/language-server-protocol/specification#textDocument_documentSymbol">{@code textDocument/documentSymbol}</a> LSP method.
+ * 
+ * <p>
+ * This interface is not meant to be injected. Use the {@link DocumentSymbolService} and the {@link HierarchicalDocumentSymbolService} directly instead.
+ * 
+ * @since 2.16
+ */
+@ImplementedBy(Noop)
+interface IDocumentSymbolService {
+
+	def List<Either<SymbolInformation, DocumentSymbol>> getSymbols(Document document, XtextResource resource,
```

Comment:
this is one of the apis where i am not sure if i like it. e.g. this could not be used to produce an eclipse outline without using lsp4j

Reply:
I am not against modifying, adjusting it, but I would like to understand what is wrong with the `lsp4j` model API? If I understood you correctly, you would like to introduce another API in `xtext-core` that would be used by `xtext-eclipse` and will be transformed into the [`IOutlineNode`](https://github.com/eclipse/xtext-eclipse/blob/master/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/outline/IOutlineNode.java), right? But `xtext-eclipse` already has `lsp4j` as a dependency, right? Then why to add another API?

Reply:
no it does not thats the point

Reply:
=> for api that might be useful toi be reused in eclipse it might help to have a lsp4j independent api e.g. as we do it for content assist

Reply:
=> we should decide that at per api level if it can be usefully used in more than the lsp4j context. currently we are inconsistent with that cc @tivervac

Reply:
=> i want at least having this thought over before introducing a new api that can be used from lsp4j only

Reply:
Also, please explain to me, what would the user do with Xtext `2.16`? Let assume, there is this new API but they already have the outline implementation in `xtext-eclipse`. What would they do then? I am obviously missing a step... That's why I cannot see the benefit of introducing this new non-`lsp4j` API.

Reply:
this is about new projects that target lsp4j and eclipse. right now they would have toi implement the outline twice

Reply:
> no it does not thats the point

So `xtext-eclipse` does not have the `lsp4j` dependency; I was not aware of it. Now I see your point. Sure, let's do it, I do not mind introducing a new API here, and the LS can map it to LSP4J and Eclipse can transform it to `IOutlineNode`s.

Reply:
see discussion at https://github.com/eclipse/xtext-eclipse/issues/711#issuecomment-419072815

Reply:
if we come to the conclusion that it is impossible to come up with a lsp4j independent api that would be ok but then i would prefer to implement it twice

Reply:
In Xtext we always allowed different levels of abstraction. We need this LSP-level. If you want something on top that abstracts over LSP, please build it. But it doesn't need to be part of this PR.

Reply:
i can live with that if we dont offer this as api to be used in eclipse

Reply:
Ok

Reply:
> i can live with that if we dont offer this as api to be used in eclipse

Should I remove the `@since 2.16` from the source? How should we proceed with this PR?

Although the PR is not super urgent, I would like to get this done if you do not have any objections. Thanks!

One thing that we have to keep in mind before we merge is the LSP4J API. See here https://github.com/eclipse/lsp4j/issues/252.

Reply:
we marked provisional api as `@Beta` before. correct @szarnekow @svenefftinge ?

Reply:
Yes, we did that in some places.